### PR TITLE
Feature/date validation on arrival

### PR DESCRIPTION
**Pull request by ho-henry-senior**
**What changes does this PR bring?**

- Arrival dates must be within one month of present date
- Applied to ui and api.
- Fixed unit tests. This involved:
    - mocking the Date object
    - updating dummy data to match
    - configuring the internationalisation object
    - some async workarounds.
- Added new tests for ui and file upload
- Refactored date within one month function into separate date parsing function

**Definition of Done**
The following need to be checked off as they are complete as part of the
PR process. If you are unable to mark off an item as complete, then state
in the comments why this is the case.
- [x] Have you built the code locally and confirmed its working?
- [ ] Is the build confirmed to be passing on CI?
- [x] Have you added enough relevant unit tests for your change?
- [ ] Have you added enough relevant E2E tests for your change?
- [x] Have you updated any relevant documentation as part of this change?
- [x] Has this change been tested against the Acceptance Criteria?
- [x] Have you considered how this will be deployed in SIT/Staging/Production?

Tested with Excel upload:
image

Tested via UI:
image

@sjose012 mentioned the following in Slack:
>DATE OF ARRIVAL in US format 4/15/2023, but it is not giving any error while uploading.

image

When I tried to enter a date in US format into the template spreadsheet, Excel gave me a data format error.

### CODE REVIEW

File: src/locales/en.json [hunk 1]
```diff
@@ -146,8 +146,9 @@
     "field_port_choice_message": "Select whether the port code is known",
     "field_departure_port_code_validation": "The departure airport code must be entered",
     "field_departure_date_validation": "Please enter Departure, today or in the future",
+    "field_arrival_date_validation": "Please enter Arrival, today or in the future",
     "field_departure_date_too_far_in_future": "Departure date cannot be more than 1 month in the future",
```
Comment (anushasivakumar) [inline src/locales/en.json:150]: Please can you change this to 'Departure date must at least be today and cannot be more than 1 month in the future'

File: src/app/api/uploadgar/validations.js [hunk 2]
```diff
@@ -48,9 +48,11 @@ module.exports.validations = (voyageObj, crewArr, passengersArr) => {
     [new ValidationRule(validator.notEmpty, '', voyageObj.departureTime, 'Enter a value for the departure time')],
     [new ValidationRule(validator.notEmpty, '', voyageObj.arrivalDate, 'Enter a value for the arrival date')],
     [new ValidationRule(validator.notEmpty, '', voyageObj.departureDate, 'Enter a value for the departure date')],
+    [new ValidationRule(validator.dateNotInPast, '', voyageObj.departureDate, 'Please enter current or future departure date')],
+    [new ValidationRule(validator.dateNotInPast, '', voyageObj.arrivalDate, 'Arrival date must atleast be today and cannot be more than 1 month in the future')],
```
Comment (anushasivakumar) [inline src/app/api/uploadgar/validations.js]: Can we reuse the same error messages as dateNotTooFarInFuture? Lines 51 and 52 could use'field_departure_date_too_far_in_future' and 'field_arrival_date_too_far_in_future'

File: src/test/api/uploadgar/post.controller.test.js [hunk 3]
```diff
@@ -250,6 +250,25 @@ describe('API upload GAR post controller', () => {
       });
     });
 
+    it('should return error if arrival date in the past', async () => {
```
Comment (anushasivakumar) [inline src/test/api/uploadgar/post.controller.test.js:253]: Nice work here !!

File: src/test/garfile/arrival/post.controller.test.js [hunk 4]
```diff
@@ -159,6 +159,27 @@ describe('Arrival Post Controller', () => {
...
+      const callController = async () => {
+        await controller(req, res);
+      };
+
+      callController().then(() => {
+        expect(garApi.get).to.have.been.calledWith('ABCDEFGH');
+        expect(garApi.patch).to.not.have.been.called;
+        expect(res.render).to.have.been.calledWith('app/garfile/arrival/index', {
+          cookie,
+          errors: [new ValidationRule(validator.currentOrFutureDate, 'arrivalDate', { d: "30", m: "5", y: "2010" }, 'Please enter Arrival, today or in the future')],
```
Comment (anushasivakumar) [inline src/test/garfile/arrival/post.controller.test.js]: Please can you check the error message

File: src/locales/en.json [hunk 5]
```diff
@@ -146,8 +146,9 @@
     "field_port_choice_message": "Select whether the port code is known",
     "field_departure_port_code_validation": "The departure airport code must be entered",
     "field_departure_date_validation": "Please enter Departure, today or in the future",
```
[the remark this answers is not in the document]
Comment (anushasivakumar) [reply src/locales/en.json:148]: Resolving as the updated file shows the changes

File: src/app/garfile/arrival/post.controller.js [hunk 6]
```diff
@@ -86,6 +87,7 @@ const buildValidations = (voyage) => {
   const validations = [
     [new ValidationRule(validator.realDate, 'arrivalDate', arriveDateObj, realDateMsg)],
     [new ValidationRule(validator.currentOrFutureDate, 'arrivalDate', arriveDateObj, futureDateMsg)],
+    [new ValidationRule(validator.dateNotTooFarInFuture, 'arrivalDate', arriveDateObj, tooFarDateMsg)],
```
Comment (anushasivakumar) [inline src/app/garfile/arrival/post.controller.js:90]: Please use 'field_arrival_date_validation' here

File: src/common/utils/validator.js [hunk 7]
```diff
@@ -224,29 +225,51 @@ function currentOrFutureDate(dObj) {
...
  */
 function dateNotTooFarInFuture(dObj) {
-  if (dObj === null || dObj === undefined) return false;
-
   const now = new Date();
-  var nextMonth = new Date(now.getFullYear(), now.getMonth() + 1, now.getDate());
+  const nextMonth = new Date(now.getFullYear(), now.getMonth() + 1, now.getDate());
+  const providedDate = getDateFromDynamicInput(dObj);
+
+  return providedDate && providedDate <= nextMonth;
```
Comment (anushasivakumar) [inline src/common/utils/validator.js:232]: This might return null instead of a boolean if the providedDate is null. Please can you check

Comment (ho-henry-senior) [reply src/common/utils/validator.js]: wrappend in Boolean()

File: src/app/garfile/arrival/post.controller.js (hunk 6 above)
Comment (ho-henry-senior) [reply src/app/garfile/arrival/post.controller.js]: done

File: src/test/garfile/arrival/post.controller.test.js (hunk 4 above)
Comment (ho-henry-senior) [reply src/test/garfile/arrival/post.controller.test.js]: have updated unit test to match code

File: src/test/api/uploadgar/post.controller.test.js [hunk 8]
```diff
@@ -250,6 +250,25 @@ describe('API upload GAR post controller', () => {
...
+      sinon.stub(XLSX, 'read').returns(data);
+
+      const callController = async () => {
+        await controller(req, res);
+      };
+
+      callController().then(() => {
+        expect(req.session.save).to.have.been.called;
+        expect(req.session.failureMsg).to.eql([
+          new ValidationRule(validator.dateNotInPast, '', '2010-07-30',  'Please enter current or future arrival date'),
```
[the remark this answers is not in the document]
Comment (ho-henry-senior) [reply src/test/api/uploadgar/post.controller.test.js]: have updated unit test to match code

File: src/test/api/uploadgar/post.controller.test.js (hunk 3 above)
Comment (ho-henry-senior) [reply src/test/api/uploadgar/post.controller.test.js:253]: think I just copied from somewhere else !!

File: src/app/api/uploadgar/validations.js (hunk 2 above)
Comment (ho-henry-senior) [reply src/app/api/uploadgar/validations.js]: done

File: src/locales/en.json (hunk 1 above)
Comment (ho-henry-senior) [reply src/locales/en.json]: done

File: src/test/garfile/departure/post.controller.test.js [hunk 9]
```diff
@@ -5,6 +5,16 @@ const sinon = require('sinon');
 const { expect } = require('chai');
 const chai = require('chai');
 const sinonChai = require('sinon-chai');
+const i18n = require('i18n');
+const path = require('path');
```
Comment (anushasivakumar) [inline src/test/garfile/departure/post.controller.test.js:9]: Thank you for fixing these tests!

**Pull request merged**